src/main/java/nextstep/security/filter/SecurityContextHolderFilter.java

```diff
@@ -0,0 +1,40 @@
+package nextstep.security.filter;
+
+import jakarta.servlet.FilterChain;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.ServletRequest;
+import jakarta.servlet.ServletResponse;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import nextstep.security.SecurityContext;
+import nextstep.security.SecurityContextHolder;
+import nextstep.security.SecurityContextRepository;
+import org.springframework.web.filter.GenericFilterBean;
+
+import java.io.IOException;
+
+public class SecurityContextHolderFilter extends GenericFilterBean {
+
+    private final SecurityContextRepository securityContextRepository;
+
+    public SecurityContextHolderFilter(SecurityContextRepository securityContextRepository) {
+        this.securityContextRepository = securityContextRepository;
+    }
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+        HttpServletRequest httpRequest = (HttpServletRequest) request;
+        HttpServletResponse httpResponse = (HttpServletResponse) response;
+
+        SecurityContext securityContext = securityContextRepository.loadContext(httpRequest);
+        SecurityContextHolder.setContext(securityContext);
+
+        try {
+            chain.doFilter(httpRequest, response);
+        } finally {
+            SecurityContext context = SecurityContextHolder.getContext();
+            securityContextRepository.saveContext(context, httpRequest, httpResponse);
+//            SecurityContextHolder.clearContext();
```

Comment:
> 4단계 SecurityContextHolderFilter 구현 중 clearContext를 SecurityContextHolderFilter에서 수행하는 게 맞다고 생각했는데, clearContext를 수행하면 테스트가 실패합니다. 일단 제출을 위해 주석처리했는데, 테스트가 실패하는 이유가 궁금합니다.

```suggestion
            securityContextRepository.saveContext(securityContext, httpRequest, httpResponse);
            SecurityContextHolder.clearContext();
```

위와 같이 수정하면 잘 동작할 것 같은데 왜 그럴지 고민해보아요 :)

Reply:
이유를 잘 모르겠네요.. ㅠㅠ 
근데 수정하다가 든 생각이 각 필터에서 `securityContextRepository.saveContext` 를 수행하고 있으니까 SecurityContextHolderFilter에서는 saveContext 수행을 안 해도 되지 않나요?